pkg/controllers/flannelmigration/migration_controller.go

```diff
@@ -0,0 +1,303 @@
+	nodeNetworkFlannel = map[string]string{migrationNodeSelectorKey: "flannel"}
+	// nodeNetworkCalico is a map value indicates a node is becoming part of Calico vxlan network.
+	// This is used both as a nodeSelector for Calico daemonset and a label for a node.
+	nodeNetworkCalico = map[string]string{migrationNodeSelectorKey: "calico"}
+	// nodeNetworkUnknown is a map value indicates a node is running network migration.
+	nodeNetworkUnknown = map[string]string{migrationNodeSelectorKey: "unknown"}
+)
+
+// flannelMigrationController implements the Controller interface.
+type flannelMigrationController struct {
+	ctx          context.Context
+	informer     cache.Controller
+	indexer      cache.Indexer
+	calicoClient client.Interface
+	k8sClientset *kubernetes.Clientset
+
+	// ipamMigrator runs ipam migration process.
+	ipamMigrator ipamMigrator
+
+	// networkMigrator runs network migration process.
+	networkMigrator *networkMigrator
+
+	// List of nodes need to be migrated.
+	flannelNodes []*v1.Node
+
+	// Configurations for migration controller.
+	config *Config
+}
+
+// NewFlannelMigrationController Constructor for Flannel migration controller
+func NewFlannelMigrationController(ctx context.Context, k8sClientset *kubernetes.Clientset, calicoClient client.Interface, cfg *Config) controller.Controller {
+	mc := &flannelMigrationController{
+		ctx:             ctx,
+		calicoClient:    calicoClient,
+		k8sClientset:    k8sClientset,
+		ipamMigrator:    NewIPAMMigrator(ctx, k8sClientset, calicoClient, cfg),
+		networkMigrator: NewNetworkMigrator(ctx, k8sClientset, calicoClient, cfg),
+		config:          cfg,
+	}
+
+	// Create a Node watcher.
+	listWatcher := cache.NewListWatchFromClient(k8sClientset.CoreV1().RESTClient(), "nodes", "", fields.Everything())
+
+	// Setup event handlers
+	handlers := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			mc.processNewNode(obj.(*v1.Node))
+			return
+		},
+	}
+
+	// Informer handles managing the watch and signals us when nodes are added.
+	mc.indexer, mc.informer = cache.NewIndexerInformer(listWatcher, &v1.Node{}, 0, handlers, cache.Indexers{})
+
+	return mc
+}
+
+// Run starts the migration controller. It does start-of-day preparation
+// and then run entire migration process. We ignore reconcilerPeriod and threadiness.
+func (c *flannelMigrationController) Run(threadiness int, reconcilerPeriod string, stopCh chan struct{}) {
+	defer uruntime.HandleCrash()
+
+	log.Info("Starting Migration controller")
+
+	// Check the status of the cluster to see if we need to migrate.
+	shouldMigrate, err := c.CheckShouldMigrate()
+	if err != nil {
+		log.WithError(err).Errorf("Error checking status, Stopping Migration controller.")
+		return
+	}
+	if !shouldMigrate {
+		log.Info("Stopping Migration controller")
+		return
+	}
+
+	// Start migration process.
+	// First step is to add node selector "projectcalico.org/node-network-during-migration==flannel" to Flannel daemonset.
```

Comment:
Before we do this, don't we need to label nodes so we don't kick flannel off every node?

Reply:
Oh, yes. That is a bug. I should add nodeSelector to Flannel daemonet after nodes get `network=Flannel` label.

Reply:
Oh, yes. That is a bug. I should add nodeSelector to Flannel daemonet after nodes get `network=Flannel` label.